File: src/lib/outreach-plan.js
const PRIORITIES = new Set(['high', 'medium', 'low']);
const SOURCES = new Set(['agent', 'demo']);

function requiredText(value, field) {
  if (typeof value !== 'string' || !value.trim()) {
    throw new TypeError(`Outreach plan ${field} is required.`);
  }
  return value.trim();
}

function normalizeItem(item) {
  if (!item || typeof item !== 'object' || Array.isArray(item)) {
    throw new TypeError('Outreach plan item must be an object.');
  }
  const priority = requiredText(item.priority, 'item priority').toLowerCase();
  if (!PRIORITIES.has(priority)) {
    throw new TypeError('Outreach plan item priority must be high, medium, or low.');
  }
  return {
    studentId: requiredText(item.student_id ?? item.studentId, 'item student id'),
    priority,
    reason: requiredText(item.reason, 'item reason'),
    suggestedNextStep: requiredText(item.suggested_next_step ?? item.suggestedNextStep, 'item suggested next step'),
  };
}

export function normalizeOutreachPlan(payload) {
  if (!payload || typeof payload !== 'object' || Array.isArray(payload)) {
    throw new TypeError('Outreach plan response must be an object.');
  }
  if (!Array.isArray(payload.items)) {
    throw new TypeError('Outreach plan items must be an array.');
  }
  const source = payload.source ?? 'demo';
  if (!SOURCES.has(source)) {
    throw new TypeError('Outreach plan source must be agent or demo.');
  }
  const generatedAt = payload.generated_at ?? payload.generatedAt ?? null;
  if (generatedAt !== null && (typeof generatedAt !== 'string' || !generatedAt.trim())) {
    throw new TypeError('Outreach plan generated time must be a string.');
  }
  return { generatedAt, source, items: payload.items.map(normalizeItem) };
}

File: backend/app/schemas/outreach_plan.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class OutreachPlanCandidate(BaseModel):
    student_id: UUID
    student_name: str
    last_contact_result: str | None = None
    last_contact_at: datetime | None = None
    open_follow_up_due_at: datetime | None = None
    teacher_confirmed_notes: list[str] = Field(default_factory=list)


class OutreachPlanItem(BaseModel):
    student_id: UUID
    priority: Literal["high", "medium", "low"]
    reason: str
    suggested_next_step: str


class OutreachPlanAgentOutput(BaseModel):
    items: list[OutreachPlanItem]


class OutreachPlanResponse(BaseModel):
    generated_at: datetime
    source: Literal["agent", "demo"]
    items: list[OutreachPlanItem]
